### PR TITLE
Plugin reload

### PR DESCRIPTION
**Pull request by gtenev**
Reloading plugin allows new versions of a plugin code to be loaded
and executed and old versions to be unloaded without restarting
the traffic server process.

More info in doc/developer-guide/plugins/reloading-plugins.en.rst

### CODE REVIEW

**Comment by gtenev** [comment]
[approve ci autest]

File: configure.ac [hunk 1]
```diff
@@ -685,6 +685,11 @@ CXX="$ac_save_CXX"
 
 TS_ADDTO(AM_CXXFLAGS, [-std=c++17])
 
+dnl find a most recent version of std::[experimental::]filesystem (c++fs or stdc++fs) available assuming c++17 and above
+AC_CHECK_LIB([c++fs], [main], ,
+    [ AC_CHECK_LIB([stdc++fs], [main], , , -lstdc++fs) ],
```
Comment (zwoop) [inline configure.ac]: I think @SolidWallOfCode tried this before, and we gave up on it.  I could not find support for this that worked universally on all our platforms.

File: doc/developer-guide/plugins/reloading-plugins.en.rst [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+PluginDso
+---------
+
+`PluginDso` - a class performing the actual DSO loading and unloading and all related initialization and
+cleanup plus related error handling. Its functionality is modularized into a separate class in hopes to
+be reused by 'global' plugins in the future.
+
+
+To make sure plugins are still loaded while their code is still in use there is reference counting done around ``PluginDso``
```
Comment (zwoop) [inline doc/developer-guide/plugins/reloading-plugins.en.rst]: Curious, why is this needed? The remap structure is ref counted now, why not piggy back the plugins to that existing ref count? Having multiple levels of ref counting seems potentially fragile no?

Comment (gtenev) [reply doc/developer-guide/plugins/reloading-plugins.en.rst]: @zwoop , as discussed at the latest ATS Summit:

-  continuations created by the plugin code, may outlive the remap structure. One example of such case would be if the plugin creates a continuation to schedule an event in the future and remap reload happens before that future moment. 
- another reason to have a separate reference counting is to use the same infrastructure code for reloading global plugin as well, in which context piggy-backing on the remap reference counting does not make much sense.

File: configure.ac (hunk 1 above)
Comment (gtenev) [reply configure.ac]: @zwoop, chatted with @SolidWallOfCode offline at the ATS summit and also looked into PR #4206 and `ts::file`. 

The idea of PR #4206 seems reasonable to me but the problem is that `ts::file` functionality is pretty limited at this moment and to use it we would need to keep enhancing it quite a bit and make sure it works on all supported platforms (i.e. this PR uses much more then just `ts::file::path`). Since in general we encourage newer C++ compilers and libraries (i.e. C+17 is minimum requirement now), I wonder if all that work is worth doing only to throw it all away soon (this was mentioned in the PR #4206 considerations as well). 

There are some functional differences in the `Filesystem` functionality between experimental and non-experimental versions (although `path /` is the only one I have noticed so far) but I wonder if we could discover and handle such differences in the our code as we go on case by case basis since the disappearance of the experimental version doesn't seem too far in the future.

Looking at our CI it seems `CentOS`, `Ubuntu`, `Debian` and `OSX` builds are fine and I was able to build on `FreeBSD 11.2` using `gcc7` (with no changes) and `clang6` (had to enable `c++experimental` by building a custom version  of `libc++`). I was curious to know what other platforms + compilers versions are failing so I can give them a try.

**Comment by gtenev** [comment]
Addressed feedback from the ATS summit: 
- plugins to be reloaded only if DSO file modification timestamp changes to avoid unnecessary DSO loads
- tricks to avoid leaving orphan runtime/temporary files and directories  in case of ATS crash

**Comment by zwoop** [comment]
Hmmm, it still fails with clang8 on FreeBSD.

**Comment by zwoop** [comment]
Doesn't build on FreeBSD12 with llvm8 :-/. I really kinda dislike all these checks for c++fs, are we sure we really want to drag in this nightmare ?

**Comment by zwoop** [comment]
Blocks #5457

**Comment by gtenev** [comment]
yeah, FreeBSD + clang seems to be problematic, will look into it

File: doc/developer-guide/plugins/reloading-plugins.en.rst [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+
+Currently only loading of "remap" plugins (`remap.config`) is supported but (re)loading of "global" plugins
+(`plugin.config`) could use same ideas and reuse some of the classes below.
+
+
+Consistent (re)loading behavior
+-------------------------------
+
+The following are some of the problems noticed during the initial experimentation:
```
Comment (vmamidi) [inline doc/developer-guide/plugins/reloading-plugins.en.rst]: I am not sure if the developer's guide is the right place for this.

Comment (zwoop) [reply doc/developer-guide/plugins/reloading-plugins.en.rst]: Agreed. If we need to retain the design documentations and reasoning, I think that should be split up? @SolidWallOfCode  Where would we put this?  In the same directory, but separate file? Or do we have a "ATS design" area somewhere ?

Comment (SolidWallOfCode) [reply doc/developer-guide/plugins/reloading-plugins.en.rst]: @mlibbey and I are supposed to be working on restructuring the current layout in the "Developer Guide" to have a coherent "Core Architecture" (or 'Internal Architecture") section. That would then be a good place to move this.

Comment (gtenev) [reply doc/developer-guide/plugins/reloading-plugins.en.rst]: checked with few people before and put it here temporarily for lack of better place, will gladly move it to where it as soon as we decide on a better place for "design + design considerations" documents. I think to keep the reasoning around will be useful for the times when we all forget what was discussed and why things were implemented the they are.

**Comment by gtenev** [comment]
[approve ci autest]

**Comment by gtenev** [comment]
[approve ci clang-analyzer]

**Comment by gtenev** [comment]
* got rid of `std::filesystem` usage (on FreeBSD `std::filesystem` is not part of llvm/clang std library by default), implemented and added what is needed/missing to `ts::file`
* moved the design document to `doc/developer-guide/design-documents`

Any other feedback, ideas, considerations ?

File: doc/developer-guide/api/functions/TSVConnCreate.en.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+Synopsis
+========
+
+`#include <ts/ts.h>`
+
+.. function:: TSCont TSVConnCreate(TSEventFunc funcp, TSMutex mutexp)
+
+Description
+===========
```
Comment (SolidWallOfCode) [inline doc/developer-guide/api/functions/TSVConnCreate.en.rst]: I'll let it slide, but perhaps a bit more verbosity in the description?

File: include/tscore/ts_file.h [hunk 5]
```diff
@@ -104,6 +104,12 @@ namespace file
     /// Get a copy of the path.
     std::string string() const;
 
+    /// Get relative path
+    self_type relative_path();
+
+    /// Get parent path
+    inline path parent_path();
```
Comment (SolidWallOfCode) [inline include/tscore/ts_file.h]: The `inline` is pointless here.

File: proxy/http/remap/PluginFactory.cc [hunk 6]
```diff
@@ -0,0 +1,268 @@
+/** @file
+
+  A brief file description
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginFactory.cc]: Sigh.

File: proxy/http/remap/PluginFactory.cc [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+/**
+ * @file PluginFactory.cc
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginFactory.cc]: Didn't you do this in line 3? Also, there's no point in putting the file name in a doxygen comment, doxygen knows what the file name is.

File: proxy/http/remap/RemapPluginInfo.cc [hunk 8]
```diff
@@ -21,60 +21,249 @@
...
+RemapPluginInfo::missingRequiredSymbolError(const std::string &pluginName, const char *required, const char *requiring)
 {
-  if (dl_handle) {
-    dlclose(dl_handle);
+  std::string error;
+  error.assign("plugin ").append(pluginName).append(" missing required function ").append(required);
+  if (requiring) {
+    error.append(" if ").append(requiring).append(" is defined");
   }
+  return error;
```
Comment (SolidWallOfCode) [inline proxy/http/remap/RemapPluginInfo.cc]: `return std::move(error);`

File: proxy/http/remap/RemapPluginInfo.cc [hunk 9]
```diff
@@ -21,60 +21,249 @@
...
+RemapPluginInfo::init(std::string &error)
+{
+  TSRemapInterface ri;
+  bool result = true;
+
+  Debug(_tag, "started initializing plugin '%s'", _configPath.c_str());
+
+  /* A buffer to get the error from the plugin instance init function, be defensive here. */
+  char tmpbuf[2048];
+  memset(tmpbuf, 0, sizeof(tmpbuf));
```
Comment (SolidWallOfCode) [inline proxy/http/remap/RemapPluginInfo.cc]: `ink_zero`?

File: doc/developer-guide/api/functions/TSVConnCreate.en.rst (hunk 4 above)
Comment (gtenev) [reply doc/developer-guide/api/functions/TSVConnCreate.en.rst]: Thanks! My goal here was not to document `TSVConnCreate()` (this is just a placeholder, I don't think I have spent enough time on this function to write proper documentation) but rather wanted to make sure all references in the documentation I wrote related to this PR feature are resolved well (without warnings).

File: proxy/http/remap/PluginFactory.cc (hunk 6 above)
Comment (gtenev) [reply proxy/http/remap/PluginFactory.cc]: Just copied the preamble somewhere from another `tscore` source file and did not notice that it had `@file` description on the top. Will merge with my comment bellow on line 24

We actually have more than 500 similar cases on master:

```
src/trafficserver $ grep -R 'A brief file description' *|wc -l
     522
```

File: proxy/http/remap/PluginFactory.cc (hunk 7 above)
Comment (gtenev) [reply proxy/http/remap/PluginFactory.cc]: Sure, will merge them (see my previous comment). 

Also will remove the name from `@file` although I do not mind specifying the `@file` name explicitly. The reason is that [the Doxygen doc about the structural commands](http://www.doxygen.nl/manual/docblocks.html#structuralcommands) feels somewhat unclear, also many of the Doxygen examples contain the `@file` name explicitly (i.e. see `structcmd.h` example right after the "Attention" note).

File: proxy/http/remap/RemapPluginInfo.cc (hunk 8 above)
Comment (gtenev) [reply proxy/http/remap/RemapPluginInfo.cc]: I wonder if this is necessary / better. 

The intent here was to rely on "named return value optimization" (NRVO [copy elision](https://en.cppreference.com/w/cpp/language/copy_elision)). 

> In a return statement, when the operand is the name of a non-volatile object with automatic storage duration, which isn't a function parameter or a catch clause parameter, and which is of the same class type (ignoring cv-qualification) as the function return type. This variant of copy elision is known as NRVO, "named return value optimization".

My understanding is that while `std::move` may be efficient, "return value optimization" typically results in faster code because the object is constructed in place and nothing is moved or copied.

File: include/tscore/ts_file.h (hunk 5 above)
Comment (gtenev) [reply include/tscore/ts_file.h]: sure, copy-and-paste remnant.

File: proxy/http/remap/RemapPluginInfo.cc (hunk 9 above)
Comment (gtenev) [reply proxy/http/remap/RemapPluginInfo.cc]: cool, changed

File: proxy/http/remap/PluginDso.cc [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+  if (_effectivePath.empty()) {
+    error.append("empty effective path");
+    result = false;
+  } else {
+    Debug(_tag, "plugin '%s' effective path: %s", _configPath.c_str(), _effectivePath.c_str());
+
+    /* Copy the installed plugin DSO to a runtime directory */
+    std::error_code ec;
+    if (!copy(_effectivePath, _runtimePath, ec)) {
+      std::string temp_error;
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginDso.cc]: Not worth changing now, but you could use `bwprint` here, e.g.
```
bwprint(error, "Failed to create a copy: {}", ec);
```

File: proxy/http/remap/PluginDso.cc [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+      temp_error.append("failed to create a copy: ").append(strerror(ec.value()));
+      error.assign(temp_error);
+      result = false;
+    } else {
+      Debug(_tag, "plugin '%s' runtime path: %s", _configPath.c_str(), _runtimePath.c_str());
+
+      /* Save the time for later checking if DSO got modified in concecutive  */
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginDso.cc]: "concecutive"?

File: proxy/http/remap/PluginFactory.cc [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+{
+  /* Discover the effective path by looking into the search dirs */
+  fs::path effectivePath = getEffectivePath(configPath);
+  if (effectivePath.empty()) {
+    error.assign("failed to find plugin '").append(configPath.string()).append("'");
+    return nullptr;
+  }
+
+  /* Only one plugin with this effective path can be loaded by a plugin factory */
+  RemapPluginInfo *plugin = (RemapPluginInfo *)findByEffectivePath(effectivePath);
```
Comment (SolidWallOfCode) [inline proxy/http/remap/PluginFactory.cc]: No C-style casts!

File: proxy/http/remap/PluginDso.cc (hunk 10 above)
Comment (gtenev) [reply proxy/http/remap/PluginDso.cc]: I see, sure, will leave it like this and consider using the suggested [internal library](https://docs.trafficserver.apache.org/en/8.0.x/developer-guide/internal-libraries/buffer-writer.en.html) in the future

File: proxy/http/remap/PluginFactory.cc (hunk 12 above)
Comment (gtenev) [reply proxy/http/remap/PluginFactory.cc]: fair enough, fixed

File: proxy/http/remap/PluginDso.cc (hunk 11 above)
Comment (gtenev) [reply proxy/http/remap/PluginDso.cc]: typo, fixed, meant `/* ... consecutive DSO reloads */`

**Comment by gtenev** [comment]
@SolidWallOfCode, could you reiterate your 👍? Pushed a change with comment clarification and c-style casting removal which removed your approving review.
